cmd/probe/internal/binding/operation_volume_protection.go

```diff
@@ -0,0 +1,460 @@
+	// the instance is running normally and there have volume(s) over the space usage threshold.
+	if !readonly && len(higher) > 0 {
+		if err := o.highWatermark(ctx, msg); err != nil {
+			return "", err
+		}
+	}
+	// the instance is protected in RO mode, and all volumes' space usage are under the threshold.
+	if readonly && len(lower) == len(o.Volumes) {
+		if err := o.lowWatermark(ctx, msg); err != nil {
+			return "", err
+		}
+	}
+	return msg, nil
+}
+
+// checkVolumeWatermark checks whether the volume's space usage is over the threshold.
+//
+//	returns 0 if the volume will not be taken in account or its space usage is under the threshold
+//	returns non-zero if the volume space usage is over the threshold
+func (o *operationVolumeProtection) checkVolumeWatermark(v volumeExt) int {
+	if v.HighWatermark == 0 { // disabled
+		return 0
+	}
+	if v.Stats.CapacityBytes == nil || v.Stats.UsedBytes == nil {
+		return 0
+	}
+	thresholdBytes := *v.Stats.CapacityBytes / 100 * uint64(v.HighWatermark)
+	if *v.Stats.UsedBytes < thresholdBytes {
+		return 0
+	}
+	return 1
+}
+
+func (o *operationVolumeProtection) highWatermark(ctx context.Context, msg string) error {
+	if o.Readonly { // double check
+		return nil
+	}
+	if err := o.lockInstance(ctx); err != nil {
+		o.Logger.Warnf("set instance to read-only error: %s, volumes: %s", err.Error(), msg)
+		return err
+	}
+
+	o.Logger.Infof("set instance to read-only OK: %s", msg)
+	o.Readonly = true
+
+	if err := o.sendEvent(ctx, reasonLock, msg); err != nil {
+		o.Logger.Warnf("send volume protection (lock) event error: %s, volumes: %s", err.Error(), msg)
+		return err
+	}
+	return nil
+}
+
+func (o *operationVolumeProtection) lowWatermark(ctx context.Context, msg string) error {
+	if !o.Readonly { // double check
+		return nil
+	}
+	if err := o.unlockInstance(ctx); err != nil {
+		o.Logger.Warnf("reset instance to read-write error: %s, volumes: %s", err.Error(), msg)
+		return err
+	}
+
+	o.Logger.Infof("reset instance to read-write OK: %s", msg)
+	o.Readonly = false
+
+	if err := o.sendEvent(ctx, reasonUnlock, msg); err != nil {
+		o.Logger.Warnf("send volume protection (unlock) event error: %s, volumes: %s", err.Error(), msg)
+		return err
+	}
+	return nil
+}
+
+func (o *operationVolumeProtection) lockInstance(ctx context.Context) error {
+	return o.BaseOperation.LockInstance(ctx)
+}
+
+func (o *operationVolumeProtection) unlockInstance(ctx context.Context) error {
+	return o.BaseOperation.UnlockInstance(ctx)
+}
+
+func (o *operationVolumeProtection) buildVolumesMsg() string {
+	volumes := make([]map[string]string, 0)
+	for _, v := range o.Volumes {
+		usage := make(map[string]string)
+		if v.HighWatermark != o.HighWatermark {
+			usage["highWatermark"] = fmt.Sprintf("%d", v.HighWatermark)
+		}
+		stats := v.Stats
+		if stats.UsedBytes == nil || stats.CapacityBytes == nil {
+			usage[v.Name] = "<nil>"
+		} else {
+			usage[v.Name] = fmt.Sprintf("%d%%", int(*stats.UsedBytes*100 / *stats.CapacityBytes))
```

Comment:
will %d get a integer overflow for stats.UsedBytes?

Reply:
The integer casting is targeted for the usage ratio, so it's safe here.